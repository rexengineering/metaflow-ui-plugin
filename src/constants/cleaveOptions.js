export const phoneNumberOptions = {
  numericOnly: true,
  delimiters: ["(", ")", " ", "-"],
  blocks: [0, 3, 0, 3, 4],
};

export const currencyOptions = {
  numeral: true,
  numeralThousandsGroupStyle: "thousand",
};

export const percentOptions = {
  numeral: true,
  numeralThousandsGroupStyle: "none",
};

export const integerOptions = {
  numeral: true,
  numeralThousandsGroupStyle: "none",
  numeralDecimalScale: 0,
};

export const floatOptions = {
  numeral: true,
  numeralThousandsGroupStyle: "none",
  numeralDecimalScale: 4,
};

export default {
  phoneNumberOptions,
  currencyOptions,
  percentOptions,
  integerOptions,
  floatOptions,
};
