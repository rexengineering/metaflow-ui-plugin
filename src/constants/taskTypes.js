import { string, number, boolean } from "yup";

// Inputs
export const TEXT = "TEXT";
export const PHONE_NUMBER = "PHONE_NUMBER";
export const PERCENTAGE = "PERCENTAGE";
export const CURRENCY = "CURRENCY";
export const BOOLEAN = "BOOLEAN";
export const INTEGER = "INTEGER";
export const FLOAT = "FLOAT";

export function isInputType(field) {
  return [
    TEXT,
    PHONE_NUMBER,
    PERCENTAGE,
    CURRENCY,
    BOOLEAN,
    INTEGER,
    FLOAT,
  ].includes(field);
}

// Info
export const TABLE = "TABLE";
export const COPY = "COPY";

export function isInfoType(type) {
  return [TABLE, COPY].includes(type);
}

const numberValidationMessage = "Please enter a number";

export const validationSchemaMapping = {
  TEXT: string("Please enter a value"),
  PHONE_NUMBER: string("Please enter a value").length(10),
  PERCENTAGE: number(numberValidationMessage).typeError(
    numberValidationMessage
  ),
  CURRENCY: number(numberValidationMessage).typeError(numberValidationMessage),
  BOOLEAN: boolean(),
  INTEGER: number(numberValidationMessage)
    .typeError(numberValidationMessage)
    .integer(),
  FLOAT: number(numberValidationMessage).typeError(numberValidationMessage),
};

export default {
  TEXT,
  PHONE_NUMBER,
  PERCENTAGE,
  CURRENCY,
  BOOLEAN,
  INTEGER,
  FLOAT,
  TABLE,
  COPY,
};
