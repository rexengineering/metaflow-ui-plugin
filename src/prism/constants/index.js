export const BUSY = "BUSY";
export const OFFLINE = "OFFLINE";
export const ONLINE = "ONLINE";

export const ActionItemDeleteType = "ActionDeleteType";
export const ActionItemCheckType = "ActionCheckType";
export const ActionItemDisplayOnlyType = "ActionDisplayOnlyType";
