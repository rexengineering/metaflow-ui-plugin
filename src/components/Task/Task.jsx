import React, { useEffect, useState, useCallback } from "react";
import {
  Button,
  CircularProgress,
  makeStyles,
  Typography,
} from "@material-ui/core";
import { object } from "yup";
import PropTypes from "prop-types";
import { Formik } from "formik";
import TaskField from "../fields/TaskField";
import {
  selectExceptionError,
  selectIsTaskBeingProcessed,
  selectIsTaskCompleted,
  selectValidationErrors,
} from "../../store/selectors";
import { completeTask } from "../../store/thunks/thunks";
import {
  convertTaskFieldsToFormUtils,
  convertValidationErrorsTo,
} from "../../utils/tasks";
import useValidateField from "../../utils/useValidateField";
import { isInfoType, isInputType } from "../../constants/taskTypes";

const useStyles = makeStyles((theme) => ({
  form: {
    display: "flex",
    flexDirection: "column",
    padding: theme.spacing(1),
  },
  submitButton: {
    margin: theme.spacing(2, 0),
  },
  field: {
    display: "inline-grid",
    marginTop: theme.spacing(2),
  },
  inProgressContainer: {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
  inProgressMessage: {
    marginLeft: theme.spacing(1),
    color: theme.palette.common.white,
  },
}));

function Task({submitButtonText, className, task, dispatch, tasksState }) {
  const { data } = task;
  const { formikInitialValues, validationSchema } =
    convertTaskFieldsToFormUtils(data);
  const [initialValues, setInitialValues] = useState(formikInitialValues ?? {});
  const formValidationSchema = object().shape(validationSchema);
  const isProcessing = selectIsTaskBeingProcessed(task, tasksState);
  const isCompleted = selectIsTaskCompleted(task, tasksState);
  const exceptionError = selectExceptionError(task, tasksState);
  const onSubmit = useCallback(
    (fields) => dispatch(completeTask(fields, task)),
    [dispatch, task]
  );
  const validateField = useValidateField(formValidationSchema);
  const classes = useStyles();
  const errors = selectValidationErrors(task, tasksState);
  const { initialErrors, initialTouched } = convertValidationErrorsTo(
    errors ?? []
  );

  useEffect(() => {
    const formUtils = convertTaskFieldsToFormUtils(data);
    setInitialValues(formUtils.formikInitialValues ?? {});
  }, [data]);

  if (isCompleted) return <Typography>Form completed!</Typography>;

  if (isProcessing || !data?.length || !initialValues) {
    return (
      <section className={classes.inProgressContainer}>
        <CircularProgress size={25} />
        <Typography className={classes.inProgressMessage} variant="body2">
          Workflow in progress
        </Typography>
      </section>
    );
  }

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={onSubmit}
      validateOnBlur={false}
      validateOnChange={false}
      enableReinitialize
      className={className}
      initialErrors={initialErrors}
      initialTouched={initialTouched}
      validationSchema={formValidationSchema}
    >
      {({ handleSubmit }) => (
        <form onSubmit={handleSubmit} className={classes.form}>
          {Array.isArray(data) &&
            data.map((field) => {
              if (isInputType(field?.type)) {
                const { dataId, label, type } = field;
                return (
                  <div key={dataId} className={classes.field}>
                    <TaskField
                      type={type}
                      id={dataId}
                      label={label}
                      validateFn={validateField}
                    />
                  </div>
                );
              }
              if (isInfoType(field?.type)) {
                const { data: fieldData, type, variant } = field;
                return (
                  <div key={data}>
                    <TaskField type={type} data={fieldData} variant={variant} />
                  </div>
                );
              }
              return <></>;
            })}
          <Button
            className={classes.submitButton}
            type="submit"
            variant="contained"
            color="secondary"
          >
            {submitButtonText}
          </Button>
          {exceptionError && (
            <Typography variant="body2" color="error">
              {exceptionError}
            </Typography>
          )}
        </form>
      )}
    </Formik>
  );
}

Task.defaultProps = {
  className: "",
  task: {},
  submitButtonText: "Submit",
};

Task.propTypes = {
  submitButtonText: PropTypes.string,
  className: PropTypes.string,
  task: PropTypes.shape({
    data: PropTypes.arrayOf(
      PropTypes.shape({
        data: PropTypes.string,
        dataId: PropTypes.string.isRequired,
        encrypted: PropTypes.bool.isRequired,
        label: PropTypes.string.isRequired,
        order: PropTypes.number.isRequired,
        type: PropTypes.string.isRequired,
        validators: PropTypes.arrayOf(
          PropTypes.shape({
            constraint: PropTypes.string,
            type: PropTypes.string.isRequired,
          })
        ),
      })
    ),
    iid: PropTypes.string,
    status: PropTypes.string,
    tid: PropTypes.string,
  }),
};

export default Task;
