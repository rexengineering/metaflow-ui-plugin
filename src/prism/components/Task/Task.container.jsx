import { connect } from 'react-redux';
import Task from "./Task";

const mapStateToProps = (state) => {
    const { rexFlow } = state;
    const { tasksState } = rexFlow;
    return {
        tasksState
    }
};

const mapDispatchToProps = (dispatch) => ({
    dispatch: dispatch,
});

export default connect(mapStateToProps, mapDispatchToProps)(Task);