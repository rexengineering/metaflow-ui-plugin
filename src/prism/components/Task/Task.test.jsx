import React from "react";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import Task from "./Task";
import getStore from "../../store";
import { TEXT } from "../../constants/taskTypes";

describe("<Task />", () => {
  const props = {
    className: "",
    task: {
      iid: "SDT32ER64W",
      status: "UP",
      tid: "TERWT65756",
      data: [
        {
          data: "Gabriel",
          dataId: "name",
          encrypted: false,
          label: "Name",
          order: 1,
          type: TEXT,
          validators: [
            {
              constraint: "",
              type: "",
            },
          ],
        },
      ],
    },
  };
  const renderTask = ({ className, task } = props, store = getStore({})) => (
    <Provider store={store}>
      <Task className={className} task={task} />
    </Provider>
  );

  it("should render", () => {
    const { asFragment } = render(renderTask());
    const initialRender = asFragment();

    expect(initialRender).not.toBeNull();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(renderTask());
    const initialRender = asFragment();

    expect(initialRender).toMatchSnapshot();
  });

  it("should render a submit button", () => {
    render(renderTask());
    const buttonNode = screen.getByText("Submit");
    expect(buttonNode).toBeInTheDocument();
  });
});
