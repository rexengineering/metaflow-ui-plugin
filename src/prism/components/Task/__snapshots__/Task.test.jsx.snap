// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Task /> should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="makeStyles-form-6"
  >
    <div
      class="makeStyles-field-8"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
          data-shrink="true"
        >
          Name
        </label>
        <div
          class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiFilledInput-input"
            name="name"
            type="text"
            value="Gabriel"
          />
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submitButton-7 MuiButton-containedSecondary"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label"
      >
        Submit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
